### PR TITLE
Prepare 240 for merge

### PR DESCRIPTION
**Pull request by trigpolynom**
This PR is a fix for #239. It builds on #240 in order to make hash-n-sign available to digital signature implementation.

See the list of [OIDs](https://github.com/IETF-Hackathon/pqc-certificates/blob/master/docs/oid_mapping.md#hash-n-sign-pqc-signatures-oids) for reference. In this PR I only implemented dilithium2 hash-n-sign algs given [this](https://github.com/open-quantum-safe/oqs-provider/issues/239#issuecomment-1748073699) comment by @baentsch.

### CODE REVIEW

**Comment by trigpolynom** [comment]
@baentsch after a lot of figuring stuff out, it seems this PR has passed all CI tests and ready for review,

**Comment by trigpolynom** [comment]
fixing the other failed checks

**Comment by trigpolynom** [comment]
@baentsch So I made the changes [here](https://github.com/open-quantum-safe/oqs-provider/pull/301/files#diff-f98b9deac5b2a64112135da6f23f87b04b5db8ae090ae7d70d2cb4598b14e9f9) to generate.py. I've tested that running generate should produce same code with passing tests. Are there other changes I need to make?

**Comment by baentsch** [comment]
> @baentsch So I made the changes [here](https://github.com/open-quantum-safe/oqs-provider/pull/301/files#diff-f98b9deac5b2a64112135da6f23f87b04b5db8ae090ae7d70d2cb4598b14e9f9) to generate.py. I've tested that running generate should produce same code with passing tests. Are there other changes I need to make?

Nope. My bad: I was expecting to see changes to the template files themselves. But you avoided that (need I assumed) by manually creating loads of new entries in the `generate.yml` file: I kind of was expecting that an orthogonal method to "mix_with" would be added, e.g., "add_hashes": That would then make easier (automate) adding hashes to all other sig algs.

So what are your plans to move this to "Ready for Review"? Iron out CI issues? Align the code points with "main" (see the comment in `generate.yml` as to which numbers should be allocated)? Anything else? More sig algs than just Dilithium2?

**Comment by trigpolynom** [comment]
@baentsch So right now I wanted to iron out the CI issues. I just fixed the asan Security Check test failure so that check should pass now. What do you think I should add? Should I do more sigalgs other than dilithium? Also I tried to change the YAML such that the hash-n-sign sigalgs would just be mixed with but I wasn't exactly sure how that should have been formatted.

**Comment by baentsch** [comment]
> Should I do more sigalgs other than dilithium? 

Yes, that would have been my suggestion (also was with the original PR): By adding (support for) more algs, the mechanisms to be added to automatically "mix in" the various digests to the sigalgs need to be created and exercized.

> Also I tried to change the YAML such that the hash-n-sign sigalgs would just be mixed with but I wasn't exactly sure how that should have been formatted

This would be defined by YAML -- should be the same syntax as with the current hybrids being "mixed in". What might be more convoluted are the changes to the generator to properly process such list of hashes-with-OIDs-and-codepoints.

But honestly, I'd leave that to you: Either one manually writes YML statements for each new alg combination to be added (you decide how many/which ones are needed) _or_ creates code to automate this for many algorithms. I typically prefer the latter as it makes `oqsprovider` more independent from upstream (e.g., name) changes -- but it's a lot of work to get right for a rare event. The second downside to full automation/matrix of digests-and-sigalg is that you need _lots of_ OIDs and code points...

**Comment by trigpolynom** [comment]
> Yes, that would have been my suggestion (also was with the original PR): By adding (support for) more algs, the mechanisms to be added to automatically "mix in" the various digests to the sigalgs need to be created and exercized.
> 
Okay cool I'll do this!

> But honestly, I'd leave that to you: Either one manually writes YML statements for each new alg combination to be added (you decide how many/which ones are needed) _or_ creates code to automate this for many algorithms. I typically prefer the latter as it makes `oqsprovider` more independent from upstream (e.g., name) changes -- but it's a lot of work to get right for a rare event. The second downside to full automation/matrix of digests-and-sigalg is that you need _lots of_ OIDs and code points...
>
And yeah this would be a lot of work. I'll first try the manual approach and see where that gets us.

**Comment by baentsch** [comment]
> Figuring this out now.

OK. FWIW, some CI failures should be fixed when you re-base to latest "main".

**Comment by trigpolynom** [comment]
@baentsch do you have any hunch on why only the above mentioned sigalgs are failing? The error code that seems to represent the cause has to do with SSL_ERROR_ZERO_RETURN

**Comment by baentsch** [comment]
I'm afraid I don't (see any "communalities"): Plain as well as Hybrid algs.... Weird. This will possibly require some serious OpenSSL debugging... I'd start with just activating OpenSSL logging. But gdb may lead to quicker insights...

**Comment by baentsch** [comment]
I'm afraid I don't (see any "communalities"): Plain as well as Hybrid algs.... Weird. This will possibly require some serious OpenSSL debugging... I'd start with just activating OpenSSL logging. But gdb may lead to quicker insights...

File: oqs-template/generate.yml
Comment (baentsch) [inline oqs-template/generate.yml:837]: This is a surefire way to cause TLS test problems: This code point is a duplicate. Please review all IDs _very_ carefully and re-test.

Comment (trigpolynom) [reply oqs-template/generate.yml:837]: Ahhhhh thank you so much for finding this!

File: test/oqs_test_hashnsign.c
Comment (baentsch) [inline test/oqs_test_hashnsign.c:75]: Why not `fprintf` (to STDERR)?

**Comment by trigpolynom** [comment]
@baentsch I definitely think the config needs to be automated and checked for duplicates. My thinking is that this should be its own PR. Which I of course would love to take on

File: oqs-template/generate.yml
Comment (baentsch) [inline oqs-template/generate.yml:408]: This is impossible: How can OID remain the same but code point change? Please explain.

File: oqs-template/generate.yml
Comment (baentsch) [inline oqs-template/generate.yml:414]: OID/code point mismatch

File: oqs-template/generate.yml
Comment (baentsch) [inline oqs-template/generate.yml:597]: Same as above

File: oqs-template/generate.yml
Comment (baentsch) [inline oqs-template/generate.yml:693]: These are deprecated algorithms -- why does anything change here?

File: oqs-template/generate.yml
Comment (baentsch) [inline oqs-template/generate.yml:856]: OID/code point mismatch

File: test/oqs_test_hashnsign.c
Comment (baentsch) [inline test/oqs_test_hashnsign.c:138]: Is this unconditional `printf` intentional? Then we cannot do a silent (successful) test run, right?

**Review by baentsch — CHANGES_REQUESTED** [review]
Thanks again for all the work that went into this @trigpolynom ! When going over it I noticed many code point changes without corresponding/matching OID changes: Why has this been done? Inadvertent? Matching up with other implementations? This needs to be clarified before merge.

Also, the sheer size (number of algorithms added) of this code change makes me nervous: Would it be conceivable to wrap this into a config variable, e.g., "OQS_ENABLE_HASHNSIGN"? By how much does the size of the provider grow with hashnsign? What's the smallest provider that can be configured/built now? This is a "nice to know" for me but possibly relevant for users, so any insight?

File: .gitignore
Comment (baentsch) [reply .gitignore:10]: If we do this, there's script adaptations needed, at least in "fullbuild.sh"

**Comment by trigpolynom** [comment]
@baentsch and @thb-sb . Thank you for these comments and suggestions, they are super helpful. I'll get to the requested changes ASAP. (Going on vaction this week, but will dive in when i get back). I'll also investigate the effect on the size of the provider that hashnsign has. Perhaps it is best if the addition of the new sigalgs is configurable and automated by a new method in generate.py?

**Comment by baentsch** [comment]
> Perhaps it is best if the addition of the new sigalgs is configurable and automated by a new method in generate.py?

Yes, this sounds like a doable approach.

> Going on vaction this week, but will dive in when i get back

Thanks for letting us know. Enjoy the holiday season then.

File: test/CMakeLists.txt
Comment (trigpolynom) [reply test/CMakeLists.txt:82]: Do we not have one for CMake files?

File: oqs-template/generate.yml
Comment (trigpolynom) [reply oqs-template/generate.yml:408]: What exactly is the relationship between OID and code point? Or what should it be?

Comment (baentsch) [reply oqs-template/generate.yml:408]: IMO it should be a one-to-one relationship as per [TLS1.3](https://www.rfc-editor.org/rfc/rfc8446): If it were not, TLS clients and servers using different algorithm implementations, but the same code point would not interoperate and nobody would understand why. This of course only is a problem until an algorithm gets standardized

File: test/CMakeLists.txt
Comment (baentsch) [reply test/CMakeLists.txt:82]: > Do we not have one for CMake files?

Not that I knew of -- pointers welcome.

**Comment by trigpolynom** [comment]
Recently picked this back up. Thinking that a good bit of refactoring needs to occur to make it configurable.

**Comment by trigpolynom** [comment]
this seems to be a good thing to monitor regarding this functionality:

https://groups.google.com/a/list.nist.gov/g/pqc-forum/c/JKMh0D0pa30/m/vbflXolxAQAJ?utm_medium=email&utm_source=footer

**Comment by baentsch** [comment]
What's your take as to how to move this PR forward, @trigpolynom ? Are you expecting input from either @thb-sb or myself? Do you want to put this into hiatus (Draft mode) until there's more clarity on the mailing list? Or even close for now?

**Comment by trigpolynom** [comment]
@baentsch Given where the conversation is in the mailing list. I would vote for closing for now.

**Comment by baentsch** [comment]
> @baentsch Given where the conversation is in the mailing list. I would vote for closing for now.

Just glancing over the discussion(s) I tend to agree... Thanks again for your work @trigpolynom and please consider submitting a new PR as and when the dust settled.

**Pull request closed without merging**
